### ISSUE TITLE
VMProtect-AntiVM Issue

### ISSUE THREAD

**Issue by 0x000007B** [hasherezade/tiny_tracer#48]
![image_No_VM](https://github.com/hasherezade/tiny_tracer/assets/143287235/cf6780e2-d7c4-4bc3-98d9-00915d78832a)
![image](https://github.com/hasherezade/tiny_tracer/assets/143287235/624911db-b098-4c1f-b5a9-f1cb44911a33)

![image](https://github.com/hasherezade/tiny_tracer/assets/143287235/3150194d-3d86-4c1e-8950-4cc5263a400d)
[VMPTest.vmp.zip](https://github.com/hasherezade/tiny_tracer/files/12446724/VMPTest.vmp.zip)  - Test executable. With the AntiVM preset enabled. VMP Demo 3.7.3, you will need to compile the source on your PC since the demo has HWID-Lock.

 
[VMPTest.vmp.exe.tag.zip](https://github.com/hasherezade/tiny_tracer/files/12446723/VMPTest.vmp.exe.tag.zip) -  Trace of the protected test executable, with the AntiVM  preset.

[VMPTest.zip](https://github.com/hasherezade/tiny_tracer/files/12446721/VMPTest.zip) - Plain EXE (No Protection)

TinyTracer.ini  ↓
 
```
ENABLE_SHORT_LOGGING=True
USE_DEBUG_SYMBOLS=False
FOLLOW_SHELLCODES=1
;FOLLOW_SHELLCODES:
; 0 : trace only the main target module
; 1 : follow only the first shellcode called from the main module
; 2 : follow also the shellcodes called recursively from the the original shellcode
; 3 : follow any shellcodes
TRACE_RDTSC=False
TRACE_INT=False
TRACE_SYSCALL=False
LOG_SECTIONS_TRANSITIONS=True
LOG_SHELLCODES_TRANSITIONS=True
HEXDUMP_SIZE=8
HOOK_SLEEP=False
SLEEP_TIME=10
; ANTIDEBUG: (Windows only)
; 0 : Disabled
; 1 : Standard
; 2 : Deep (may lead to some false positives)
ANTIDEBUG=1
ANTIVM=0
```

There is an issue with a tiny tracer currently related to VMProtect. The issue is that when AntiVM is enabled and under the tiny tracer, VMProtect throws an error that says "Sorry this application cannot run under a Virtual Machine.", I'm not on a VM. So I don't know why it's throwing that error while running under a tiny tracer. I suppose the newest tiny tracer update has something to do with this happening. Based on the "Trap Flag". But I'm not so sure.

Regards Dynamic.

**Comment by hasherezade** [comment]
hi @0x000007B !
Thanks for reporting, I will check it soon (not at home right now). But the first thing that I noticed is that you are not using the latest TinyTracer (which is [2.7](https://github.com/hasherezade/tiny_tracer/releases/tag/2.7)). Can you try first with the latest one?

**Comment by hasherezade** [comment]
Also noticed that the case *without* "Virtualization tools" runs ok. And the recently implemented bypass was for  "Usermode + kernelmode debugger" - so this works fine. There is nothing wrong with the "Trap Flag" bypass.

Just the bypass for the mode *with* "Virtualization tools" is not implemented yet. I will see what exactly it checks for and implement this bypass in the further release. Probably it is about the presence of the injected DLL, or maybe the altered RDTSC.

**Comment by 0x000007B** [comment]
Hey, @hasherezade !, thanks for the quick reply. Yes my theory was wrong, I thought the "Trap Flag" played in this error. But I was wrong sorry. I now understand, Thanks!

**Comment by hasherezade** [comment]
Hi @0x000007B ! I recently had more time to dig into this issue, and it turns out that you was right, and I was wrong 😅.
I thought that it can't be the problem with the Trap Flag, because I already implemented the emulation, and it was bypassing the Trap Flag test in Al-Khaser, and some others. But it turns out that the checks done by VMProtect are more detailed (finally reading the code of leaked VMProtect gave me more clarity). So I realized that my implementation of Trap Flag emulation was inaccurate, and VMProtect was still catching it!
I fixed the implementation, and it should work fine now. Those are the settings in the TinyTracer.ini with which I tested:

```ini
[...]
ANTIDEBUG=1
;ANTIDEBUG: (Windows only)
; 0 : Disabled
; 1 : Standard
; 2 : Deep (may lead to some false positives)
ANTIVM=1
;ANTIVM: (Windows only)
; 0 : Disabled
; 1 : Standard
; Settings for ANTIVM enabled:
EMULATE_HYPERV=False
; Settings for ANTIDEBUG enabled:
EMULATE_SINGLE_STEP=True
;EMULATE_SINGLE_STEP:
; On True: when the trap flag was set, throw the SINGLE_STEP exception
; On False: the trap flag will be removed and ignored (no exception)
```

I don't know if you still need it, but I thought to clarify it. Sorry for getting it wrong last time! I was busy and I didn't dig deep enough in what is really going on.
